gpflow/experimental/check_shapes/checker.py

```diff
@@ -49,17 +50,93 @@ def _as_parsed_tensor_spec(tensor_spec: TensorSpecLike, context: ErrorContext) -
         if isinstance(tensor_spec, tuple):
             for dim in tensor_spec:
                 if isinstance(dim, int):
-                    dimension_specs.append(ParsedDimensionSpec(dim, None, False))
+                    dimension_specs.append(ParsedDimensionSpec(dim, None, False, False))
                 else:
                     assert dim is None
-                    dimension_specs.append(ParsedDimensionSpec(None, None, False))
+                    dimension_specs.append(ParsedDimensionSpec(None, None, False, False))
         else:
             assert tensor_spec is None
-            dimension_specs.append(ParsedDimensionSpec(None, None, True))
+            dimension_specs.append(ParsedDimensionSpec(None, None, True, False))
         shape = ParsedShapeSpec(tuple(dimension_specs))
         return ParsedTensorSpec(shape, None)
 
 
+@dataclass
+class _ObservedDim:
+    """
+    Storage of observed size of a single dimension variable.
+    """
+
+    size: Optional[int] = None
+
+    def update(self, actual: Optional[int], broadcast: bool) -> bool:
+        """
+        Update this observation with new data.
```

Comment:
Only "updates" if the `self.size` is `None` - not sure I'm following what this method is supposed to do?

Reply:
> Only "updates" if the `self.size` is `None` - not sure I'm following what this method is supposed to do?

I've rewritten the documentation.